src/ethereum/constantinople/state.py

```diff
@@ -0,0 +1,499 @@
+
+
+def commit_transaction(state: State) -> None:
+    """
+    Commit a state transaction.
+
+    Parameters
+    ----------
+    state : State
+        The state.
+    """
+    state._snapshots.pop()
+
+
+def rollback_transaction(state: State) -> None:
+    """
+    Rollback a state transaction, resetting the state to the point when the
+    corresponding `start_transaction()` call was made.
+
+    Parameters
+    ----------
+    state : State
+        The state.
+    """
+    state._main_trie, state._storage_tries = state._snapshots.pop()
+
+
+def get_account(state: State, address: Address) -> Account:
+    """
+    Get the `Account` object at an address. Returns `EMPTY_ACCOUNT` if there
+    is no account at the address.
+
+    Use `get_account_optional()` if you care about the difference between a
+    non-existent account and `EMPTY_ACCOUNT`.
+
+    Parameters
+    ----------
+    state: `State`
+        The state
+    address : `Address`
+        Address to lookup.
+
+    Returns
+    -------
+    account : `Account`
+        Account at address.
+    """
+    account = trie_get(state._main_trie, address)
+    if isinstance(account, Account):
+        return account
+    else:
+        return EMPTY_ACCOUNT
+
+
+def get_account_optional(state: State, address: Address) -> Optional[Account]:
+    """
+    Get the `Account` object at an address. Returns `None` (rather than
+    `EMPTY_ACCOUNT`) if there is no account at the address.
+
+    Parameters
+    ----------
+    state: `State`
+        The state
+    address : `Address`
+        Address to lookup.
+
+    Returns
+    -------
+    account : `Account`
+        Account at address.
+    """
+    account = trie_get(state._main_trie, address)
+    return account
+
+
+def set_account(
+    state: State, address: Address, account: Optional[Account]
+) -> None:
+    """
+    Set the `Account` object at an address. Setting to `None` deletes
+    the account (but not its storage, see `destroy_account()`).
+
+    Parameters
+    ----------
+    state: `State`
+        The state
+    address : `Address`
+        Address to set.
+    account : `Account`
+        Account to set at address.
+    """
+    trie_set(state._main_trie, address, account)
+
+
+def destroy_account(state: State, address: Address) -> None:
+    """
+    Completely remove the account at `address` and all of its storage.
+
+    This function is made available exclusively for the `SELFDESTRUCT`
+    opcode. It is expected that `SELFDESTRUCT` will be disabled in a future
+    hardfork and this function will be removed.
+
+    Parameters
+    ----------
+    state: `State`
+        The state
+    address : `Address`
+        Address of account to destroy.
+    """
+    if address in state._storage_tries:
+        del state._storage_tries[address]
+    set_account(state, address, None)
+
+
+def destroy_storage(state: State, address: Address) -> None:
+    """
+    Completely remove the account at `address` and all of its storage.
```

Comment:
@Bhargavasomu

Reply:
The main functionality here is to destroy the storage instead of destroying the account. I just introduced this function to insist that we are deleting the storage instead of the whole account. I guess the docstring here is wrong and will change this to be more meaningful of the function name.